File: matlab/collisions/generateTargets.m
function output=generateTargets(density,bench)
% usage output=generate_targets(density,bench)
% density is the # targets per patrol area
% specify bench if you have one, otherwise a full bench will be generated
%
% generate a uniform density set of targets over the field of view
% of the bench
%
% the following will generate an as-is bench from the CfgXml file:
% bench = defineBenchGeometry([],1,1)


toggle.makefigs = true;


% bench definition
if ~exist('bench','var')
    centers = getCentersRails(14);
    centers = bsxfun(@times, [1 exp(i*2*pi/3) exp(i*4*pi/3)], centers);
    centers = reshape(centers,[],1);
    bench = defineBenchGeometry(centers,1,1);
    bench.alpha = 0;
    clear centers
end
ncobras = length(bench.center);

%% uniformly generate targets over field of view.
ntgt = ceil(density * (bench.field.R/median(bench.rMax))^2);
% $$$ fprintf(1,'Generating %d targets for %d positioners\n',ntgt,ncobras);

THT = rand(1,ntgt) * 2 * pi;
RR  = sqrt(rand(1,ntgt)) * bench.field.R;
tgt = RR .* exp(i*THT); % generated targets, centered on cobra CM
tgt = tgt + bench.field.cm;

output = tgt;

File: python/ics/cobraOps/generateTargets.py

import numpy as np


def generateTargets(density, bench, makefigs=True):
    """Generate a uniform density set of targets over the field of view of the bench.

    Parameters
    ----------
    density: int
        The # targets per patrol area.
    bench: object
        The bench to use. If None, a full bench will be generated.
        Default is None.
    makefigs: bool
        If true, some figures will be made. Default is True.
    
    Returns
    -------
    Object
        The target list.
            
    """
    # Define the bench if necessary
    if bench is None:
        # Get the first sector
        firstSector = getCentersRails(14)
            
        # Create the centers array
        cobrasPerSector = len(firstSector)
        centers = np.zeros(3 * cobrasPerSector, dtype="complex")
            
        # Add the first sector
        centers[:cobrasPerSector] = firstSector 
            
        # Add the second sector rotating the first sector 120 degrees
        centers[cobrasPerSector:2 * cobrasPerSector] = firstSector * np.exp(1j * 2 * np.pi / 3) 

        # Add the third sector rotating the first sector 240 degrees
        centers[2 * cobrasPerSector:] = firstSector * np.exp(1j * 4 * np.pi / 3)

        bench = defineBenchGeometry(centers, True, True)
        bench["alpha"] = 0
    
    ncobras = len(bench["center"])

    # uniformly generate targets over field of view.
    ntgt = int(np.ceil(density * (bench["field"]["R"] / np.median(bench["rMax"])) ** 2))
    print("Generating " + str(ntgt) + " targets for " + str(ncobras) + " positioners")

    THT = np.random.random(ntgt) * 2 * np.pi
    RR = np.sqrt(np.random.random(ntgt)) * bench["field"]["R"]
    tgt = RR * np.exp(1j * THT)  # generated targets, centered on cobra CM
    tgt += bench["field"]["cm"]

    return tgt
